proposals/mqtt.md

```diff
@@ -0,0 +1,83 @@
+# [MQTT](https://mqtt.org/) Discovery Handler
+
+## MQTT and its use
+
+MQTT is a simple message bus with a broad use in IoT devices, such device will usually connect to an existing MQTT broker on the network.
+This allows reuse of the same broker for multiple devices.
+
+Let's quickly introduce some key concepts of MQTT:
+
+- An MQTT message consists of a slash (`/`) separated UTF-8 topic and a completely free-form payload with no built-in notion of sender
+- Topics are only used for message routing and are not "created" or "destroyed"
+- By default, messages are not stored, and a message with no client subscribing to its topic is simply discarded
+- There are no specified way of querying connected clients or current subscriptions
+- An MQTT client can use wildcards when subscribing to topics
+
+Because of these properties, it is impossible to discover devices that do not publish any message.
+
+Usually a device will publish into topics that are specific to that device instance so that it becomes possible to know the origin of the message.
+For example a weather station might publish into topics like `weather/<city>/temperature` and `weather/<city>/pressure` with `<city>` the location of station.
+
+Many power-constrained devices are not maintaining a long-lived connection to the MQTT broker.
+
+In the rest of the document we assume that devices to discover follow this kind of pattern.
+
+Using the above description of MQTT and device behavior, if we consider the Discovery Handler to be an MQTT client, we can define a discoverable device as an MQTT client that will "frequently" publish to an MQTT topic.
+
+## Discovery Details and Properties
+
+In order to discover MQTT based devices, we need the following pieces of information:
+
+- The MQTT broker to connect to (and its credentials)
+- The topic patterns to subscribe to
+- How long shall we wait to consider a silent device as disconnected
+
+Here is the expected schema for the discovery details:
+
+```yaml
+discoveryDetails:
+    mqttBrokerUri:
+        type: string
+    topics:
+        type: array
+        items:
+            type: string
+    timeoutSeconds: 
+        type: int64
+        minimum: 0
+        exclusiveMinimum: true
+```
+
+If the following keys are present in the discovery properties they will be used for connection and authentication to the MQTT broker:
+
+- `mqtt_broker_username`: username for authentication against the broker
+- `mqtt_broker_password`: password for authentication against the broker
+- `mqtt_broker_certificate`: client TLS certificate for authentication against the broker
+- `mqtt_broker_key`: client TLS key for authentication against the broker
+- `mqtt_broker_ca`: TLS CA bundle for validating the broker TLS certificate (if unspecified, system-wide bundle is used)
+
+## Instance properties
+
+The ID of the device is the full topic used by the device.
```

Comment:
what if a device publishes multiple topics? Do we create multiple instances for it?

Reply:
Yes, the base idea here is to craft the query to only match a channel per device